modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala

```diff
@@ -1445,14 +1480,21 @@ private[smithy4s] class SchemaVisitorJCodec(
           cursor: Cursor,
           in: JsonReader
       ): scala.collection.Map[String, Any] => Z = {
+        val unknownValues =
+          if (keepUnknown) ListBuffer[(String, Document)]() else null
```

Comment:
I appreciate the effort to minimise allocations :+1: . I would personally have gone for distinct `JCodec` implementations based on whether the structure has a `@jsonUnknown` field or not, which does induce some copy/pasting, but completely shaves off some checks on the critical path.

Reply:
There are now 2 `JCodec` implementations to separately handle structs with a `jsonUnknown` member and ones without.